### PR TITLE
Internal eventstream RPC functionality to support greengrass RPC

### PR DESCRIPTION
**Pull request by bretambrose**
Eventstream RPC implementation
Testing is done via the (currently handwritten) echo rpc client and tests, which are in a separate PR to keep things managable
Will not build until the CRT is released with https://github.com/awslabs/aws-crt-nodejs/pull/439

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: lib/eventstream_rpc_utils.ts [hunk 1]
```diff
@@ -0,0 +1,633 @@
...
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+import * as eventstream_rpc from "./eventstream_rpc";
+import {CrtError, eventstream} from "aws-crt";
+
+/*
+ * Internal utility functions for generated RPC clients to perform normalization, serialization, deserialization, and
+ * validation.
```
Comment (TwistedTwigleg) [inline lib/eventstream_rpc_utils.ts:11]: Just to make sure my understanding is correct: These functions are not necessarily being used right now, but they will be used in the generated RPC clients?

File: lib/eventstream_rpc_utils.ts [hunk 2]
```diff
@@ -0,0 +1,633 @@
...
+ * Transforms an eventstream payload type (an opaque blob) to a base64-encoded string
+ *
+ * @param payload blob to transform
+ * @return a base64-encoded string
+ */
+export function encodePayloadAsString(payload : eventstream.Payload) : string {
+    if (typeof payload === "string") {
+        return Buffer.from(payload).toString("base64");
+    } else {
+        return Buffer.from(payload).toString("base64");
```
Comment (TwistedTwigleg) [inline lib/eventstream_rpc_utils.ts:26]: Nitpick: Use `return Buffer.from(payload).toString("base64")`?

File: lib/eventstream_rpc.ts [hunk 3]
```diff
@@ -0,0 +1,1153 @@
...
+    DeserializationError,
+
+    /**
+     * An error occurred during the connect-connack handshake between client and server.  Usually this means
+     * the connect was not accepted by the server and thus hints at an authentication problem.
+     */
+    HandshakeError,
+
+    /**
+     * An error that isn't classifiable occurred.
```
Comment (TwistedTwigleg) [inline lib/eventstream_rpc.ts:107]: Nitpick: change to `... isn't classifiable by one of the other RpcErrorTypes occurred`?

File: lib/eventstream_rpc.ts [hunk 4]
```diff
@@ -0,0 +1,1153 @@
...
+     */
+    ClientStateError,
+
+    /**
+     * An error occurred ostensibly due to an underlying networking failure.
+     */
+    NetworkError,
+
+    /**
+     * An error that occurs when the underlying transport is shut down before an expected protocol event occurs.
```
Comment (TwistedTwigleg) [inline lib/eventstream_rpc.ts:123]: Nitpick: The format of this comment is slightly different. I'd change it to `Underlying transport is shutdown before an excepted protocol event occurs` or `An error occurred due to the underlying transport shutdown before an excepted protocol event occurs`.

File: lib/eventstream_rpc.ts [hunk 5]
```diff
@@ -0,0 +1,1153 @@
...
+export interface RpcMessageTransformationOptions {
+
+    /**
+     * (CONNECT) message to transform
+     */
+    message: eventstream.Message,
+
+    /**
+     * Optional controller that allows for cancellation of the asynchronous process.  The transformation implementation
+     * is responsible for respecting this.
```
Comment (TwistedTwigleg) [inline lib/eventstream_rpc.ts:223]: I think I know what you're saying here, but just to confirm: You mean that the transformation implementation is responsible for canceling the async process, correct? The canceling can only happen in the transformation?

File: lib/eventstream_rpc.ts [hunk 6]
```diff
@@ -0,0 +1,1153 @@
...
+
+/**
+ * Utility function to create RpcError errors
+ *
+ * @param type type of error
+ * @param description longer description of error
+ * @param internalError optional CrtError that caused this error
+ * @param serviceError optional modeled eventstream RPC service error that triggered this error
+ */
+export function createRpcError(type: RpcErrorType, description: string, internalError?: CrtError, serviceError?: any) {
```
Comment (TwistedTwigleg) [inline lib/eventstream_rpc.ts:998]: Nitpick: Missing `@return`?

**Review by TwistedTwigleg — APPROVED** [review]
Looks good to me! There was just a few minor nitpicks.

File: lib/eventstream_rpc.ts [hunk 7]
```diff
@@ -0,0 +1,1153 @@
...
+/**
+ * @internal a rough mirror of the internal connection state, but ultimately must be independent due to the more
+ * complex connection establishment process (connect/connack).  Used to prevent API invocations when the client
+ * is not in the proper state to attempt them.
+ */
+enum ClientState {
+    None,
+    Connecting,
+    Connected,
+    Finished,
```
[the remark this answers is not in the document]
Comment (bretambrose) [reply lib/eventstream_rpc.ts:275]: Stream ending does not necessarily imply  anything about underlying connection status.

File: lib/eventstream_rpc.ts [hunk 8]
```diff
@@ -0,0 +1,1153 @@
...
+
+/**
+ * Event listener type signature for listening to ended events for streaming operations
+ */
+export type StreamingOperationEndedListener = (eventData: StreamingOperationEndedEvent) => void;
+
+/**
+ * Implementation of a bi-direction streaming operation.
+ *
+ * TODO: may change slightly for uni-directional operations
```
[the remark this answers is not in the document]
Comment (bretambrose) [reply lib/eventstream_rpc.ts:839]: This isn't final.

File: lib/eventstream_rpc.ts [hunk 9]
```diff
@@ -0,0 +1,1153 @@
...
+    /**
+     * Optional cancel controller to cancel the sending of eventstream messages with.  Cancellation includes both
+     * operation activation and sending of streaming messages.  It does not affect a streaming operation's state.
+     */
+    cancelController?: cancel.ICancelController;
+
+    /**
+     * Disables client-side data validation.  Useful for testing how the client handles errors from the service.
+     */
+    disableValidation? : boolean;
```
[the remark this answers is not in the document]
Comment (bretambrose) [reply lib/eventstream_rpc.ts:623]: I don't have an opinion one way or another about disabling validation.

File: lib/eventstream_rpc.ts (hunk 5 above)
Comment (bretambrose) [reply lib/eventstream_rpc.ts:223]: Yes

File: lib/eventstream_rpc.ts [hunk 10]
```diff
@@ -0,0 +1,1153 @@
...
+    /**
+     * A formally-modeled error sent from server to client
+     */
+    ServiceError
+}
+
+/**
+ * @internal
+ */
+interface RpcErrorModel {
```
[the remark this answers is not in the document]
Comment (bretambrose) [reply lib/eventstream_rpc.ts:141]: It's not user facing

File: lib/eventstream_rpc_utils.ts (hunk 1 above)
Comment (bretambrose) [reply lib/eventstream_rpc_utils.ts:11]: Yes

File: lib/eventstream_rpc.ts [hunk 11]
```diff
@@ -0,0 +1,1153 @@
...
+                reject(createRpcError(RpcErrorType.HandshakeError, "Failed to establish eventstream RPC connection - invalid connack"));
+                setImmediate(() => { this.close(); });
+                return;
+            }
+
+            /*
+             * Remove the promise-rejecting disconnect listener and replace it with a regular old listener that
+             * doesn't reject the connect() promise since we're going to resolve it now.
+             */
+            this.connection.removeListener('disconnection', onDisconnectWhileConnecting);
```
[the remark this answers is not in the document]
Comment (bretambrose) [reply lib/eventstream_rpc.ts:434]: Correct

**Pull request merged**